### PR TITLE
Improve performance of interpolation in decalarative sources

### PR DESCRIPTION
**Pull request by szubster**
## What
`interpolated_string.py` is called thousands or even millions of times in declarative streams.
E.g. in `datetime_based_cursor.py`.
It is often the case (e.g. in jira source) that there is no evaluation needed, and input is just a string.
When encountered with multiple parent streams this can lead to jinja evaluation taking most of the CPU time.

Jinja compiled templates are also compiled on each evaluation. Which makes them slow.

## How
Check whether evaluated string equals input on first try. If so, use input string directly.
Add cache for compiled jinja templates.

## Review guide
1. `interpolated_string.py`

## User Impact
Improved performance.
Example of small run on our jira instance (tests from changes in `interpolated_string.py` only:
pre:
```
Executed in  267.01 secs    fish           external
   usr time  153.66 secs    0.11 millis  153.66 secs
   sys time    1.10 secs    1.35 millis    1.10 secs
```
post:
```
Executed in  130.13 secs    fish           external
   usr time   24.47 secs    0.18 millis   24.47 secs
   sys time    0.79 secs    2.05 millis    0.78 secs
```

Difference in test execution for whole interpolated package (with jinja improvements):

| | before | after |
| --- | --- | ---|
| single eval | 0.75s | 0.76s |
| loop 100 evals on single object | 14.93s | 1.36s |

## Can this PR be safely reverted and rolled back?

- [x] YES 💚
- [ ] NO ❌

### CODE REVIEW

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/interpolation/interpolated_string.py
Comment (natikgadzhi) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/interpolation/interpolated_string.py:31]:
```suggestion
        self._is_plain_string = None
```

Naming doesn't sit right with me — something like this perhaps? Up to @artem1205 and @girarda though.

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/interpolation/interpolated_string.py
Comment (natikgadzhi) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/interpolation/interpolated_string.py:41]: Since this does a bit of caching magic, I'd love a comment explaining why you're doing this in eval specifically. I understand you need config to actually interpolate, so this is not insane, but a comment for posterity would be good.

Also, are there scenarios where the first interpolation would render the same string, but later interpolation would render something different? I can't think of it, but is there a world where a jinja expression renders itself / morphs conditionally on an input?

**Review by natikgadzhi — COMMENTED** [review]
This is a huge gain 👏

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/interpolation/interpolated_string.py
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/interpolation/interpolated_string.py:41]: +1. The change makes sense to me, but let's describe the rationale for posterity.

That being said, thank you for taking the time to contribute to Airbyte @szubster ! <3

**Comment by szubster** [comment]
Thank you for the comments.
I think I have found few more tricks to speed whole jinja interpolation. Not just the simple case.
Give me some time and I will come back to this.

**Comment by szubster** [comment]
Is there some synthetic benchmark or a test suite I could use to test those changes?

**Comment by girarda** [comment]
@szubster there aren't performance benchmark tests, but depending on where your follow up changes land:
- [there's some tests specific to InterpolatedString](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/unit_tests/sources/declarative/interpolation/test_interpolated_string.py)
- [there's some tests for the underlying interpolation logic](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/unit_tests/sources/declarative/interpolation/test_jinja.py)

We'll also run regression tests with a few connectors to make sure this doesn't introduce regressions

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/interpolation/interpolated_string.py
Comment (szubster) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/interpolation/interpolated_string.py]: If there is any (which I doubt) then the output would still not look like input?
There are code-challenges, like https://en.wikipedia.org/wiki/Quine_(computing) but I doubt it's a real-life scenario.
And even if it is, then the logic is still correct :)

**Comment by szubster** [comment]
@natikgadzhi @girarda Please check again. Added more performance improvements.

**Comment by szubster** [comment]
I am eagerly waiting for performance numbers from your internal regression tests :)

**Comment by natikgadzhi** [comment]
Kicked off the CI again. I personally like it, but the ultimate decision to merge is for @girarda.

**Review by natikgadzhi — APPROVED** [review]

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/interpolation/interpolated_string.py
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/interpolation/interpolated_string.py:32]:
```suggestion
        # This allows for optimization, but we do not know it yet at this stage
```

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/interpolation/jinja.py
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/interpolation/jinja.py:134]:
```suggestion
        We must cache the Jinja Template ourselves because we're using `from_string` instead of a template loader
```

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/interpolation/interpolated_string.py
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/interpolation/interpolated_string.py:47]:
```suggestion
            # This indicates occurence of a plain string, not a template and we can skip Jinja in subsequent runs.
```

**Review by girarda — APPROVED** [review]
I'm a fan! Thanks @szubster <3

**Comment by szubster** [comment]
Checked test results. It is a bit more nuanced...
ValueError should be raised only when for top-level variables in jinja context.
When there is a dict in top-level but does not contain a specific key, `''` should be returned

**Comment by szubster** [comment]
All checks are passing now :)

**Comment by girarda** [comment]
Running regression tests for a few connectors. I'll merge after they run successfully 🎉 
- https://github.com/airbytehq/airbyte/actions/runs/10496661776
- https://github.com/airbytehq/airbyte/actions/runs/10496674783
- https://github.com/airbytehq/airbyte/actions/runs/10496688648

**Comment by szubster** [comment]
There is one failure, but I do not see how it could be related to my change?
`Error: Invalid value for '--name': 'source-quickbook' in ....`

**Comment by natikgadzhi** [comment]
It should be quickbooks — must be a typo ;)

**Comment by girarda** [comment]
@szubster this is now available in CDK 4.5.3 🎉 

Thank you for the huge improvement <3

**Comment by szubster** [comment]
If you would have any numbers from synthetic tests or anything to share how it improved performance, I would be glad to hear :)

**Pull request merged**
